File: lib/compiler/emitter.dart
import 'package:bliss/compiler/macro_expander.dart';
import 'package:bliss/compiler/node.dart';
import 'package:bliss/utils.dart';

const _operators = {
  /* Math */
  "+": "+",
  "-": "-",
  "*": "*",
  "/": "/",
  /* Equality */
  "<": "<",
  "<=": "<=",
  ">": ">",
  ">=": ">=",
  "=": "==",
  "not=": "!=",
  /* Logical */
  "and": "&&",
  "or": "||",
};

const moduleSeparator = "#";

const _specialFormsNotTerminatedWithSemicolons = {"do"};

class Emitter {
  /*================================== Data ==================================*/

  /*========================= General utilities etc. =========================*/

  Future<String> _tryFormatIr(String ir) async {
    // Try to format using both refmt and bsrefmt

    final (
      bsrefmtRes,
      bsrefmtExitCode,
    ) = await runCommandWithStdinAndFallbackValueOnError(
      "bsrefmt",
      [],
      valueOnError: ir,
      stdin: [ir],
    );

    if (bsrefmtExitCode == 0) return bsrefmtRes;

    final (
      refmtRres,
      refmtExitCode,
    ) = await runCommandWithStdinAndFallbackValueOnError(
      "refmt",
      [],
      valueOnError: ir,
      stdin: [ir],
    );

    if (refmtExitCode == 0) return refmtRres;

    return ir;
  }

  /*======================= Helper emission functions ========================*/

  String _emitCommaSeparatedItems(List<Node> itemNodes) {
    return itemNodes.map(_emitExpr).join(", ");
  }

  String _emitStatementBody(List<Node> bodyNodes) {
    return bodyNodes
        .map((node) {
          final exprStr = _emitExpr(node);

          final id =
              node.contents is List && node.contents.isNotEmpty
                  ? ((node.contents[0] as SymbolNode).contents as String)
                  : "";
          final shouldTerminateWithSemicolon =
              !_specialFormsNotTerminatedWithSemicolons.contains(id);

          return "$exprStr${shouldTerminateWithSemicolon ? ";" : ""}";
        })
        .join("\n");
  }

  /*==================== Special form (keyword) emissions ====================*/

  String _emitModuleDefinition(List<Node> nodes) {
    switch (nodes) {
      case [final SymbolNode name, ...final bodyNodes]:
        final nameStr = _emitSymbol(name);
        final bodyStr = _emitStatementBody(bodyNodes);
        return "module $nameStr = {\n$bodyStr\n}";
      default:
        throw "Invalid module definition $nodes";
    }
  }

  String _emitModuleImport(List<Node> nodes) {
    switch (nodes) {
      // (use List (sort (cons 0 [1 2 3])))
      case [final SymbolNode name, final ListNode bodyNode]:
        final nameStr = _emitSymbol(name);
        final bodyStr = _emitList(bodyNode);
        return "$nameStr.($bodyStr)";
      // List.(sort(cons(0, [1, 2, 3])))
      // (use List)
      case [final SymbolNode name]:
        final nameStr = _emitSymbol(name);
        return "open $nameStr";
      default:
        throw "Invalid module import $nodes";
    }
  }

  String _emitFunctionDefinition(
    List<Node> nodes, {
    bool private = false,
    bool lambda = false,
    bool recursive = false,
  }) {
    String emitFunction({
      SymbolNode? name,
      List<Node>? args,
      required List<Node> bodyNodes,
    }) {
      switch (bodyNodes) {
        default:
          final nameStr =
              name != null ? _emitSymbol(name, private: private) : "";
          final argsStr = args != null ? _emitCommaSeparatedItems(args) : "";
          final bodyStr = _emitStatementBody(bodyNodes);
          final maybeRecursiveStr = !lambda && recursive ? " rec " : " ";
          final keywordStr = !lambda ? "let$maybeRecursiveStr$nameStr = " : "";
          return "$keywordStr($argsStr) => {\n$bodyStr\n}";
      }
    }

    return switch (nodes) {
      // Regular function
      // (def[p] run [...] ...) (regular function with arguments etc.)
      [
        final SymbolNode name,
        ValueListNode(contents: final args as List<Node>),
        ...final bodyNodes,
      ]
          when args.every((x) => x is SymbolNode) =>
        emitFunction(name: name, args: args, bodyNodes: bodyNodes),
      // (def[p] run ...) (shorthand for a function with no arguments)
      [final SymbolNode name, ...final bodyNodes] => emitFunction(
        name: name,
        bodyNodes: bodyNodes,
      ),
      // Anonymous/lambda function
      // (fn [...] ...) (anonymous/lambda function with arguments etc.)
      [ValueListNode(contents: final args as List<Node>), ...final bodyNodes]
          when lambda && args.every((x) => x is SymbolNode) =>
        emitFunction(args: args, bodyNodes: bodyNodes),
      // (fn ...) (shorthand for an anonymous/lambda with no arguments)
      [...final bodyNodes] when lambda => emitFunction(bodyNodes: bodyNodes),
      _ => throw "Invalid function declaration $nodes",
    };
  }

  String _emitVariable(
    List<Node> nodes, {
    bool mutable = false,
    bool private = false,
  }) {
    switch (nodes) {
      case [final Node name, final value] when name is! ListNode:
        final nameStr = _emitExpr(name);
        final valueStr = _emitExpr(value);
        final maybeMutableValueStr = mutable ? "ref($valueStr)" : valueStr;
        return "let $nameStr = $maybeMutableValueStr";
      default:
        throw "Invalid variable declaration $nodes";
    }
  }

  String _emitVariableMutation(List<Node> nodes) {
    switch (nodes) {
      case [final name, final value]:
        void addCaretToVariableNameInValue(List<Node> valueNodes, String name) {
          for (final node in valueNodes) {
            switch (node) {
              case ListNode(contents: final nodes) ||
                  ValueListNode(contents: final nodes):
                addCaretToVariableNameInValue(nodes, name);
              case SymbolNode(contents: final String str) when str == name:
                value.contents = "$str^";
              default:
                null;
            }
          }
        }

        final nameStr = _emitExpr(name);
        addCaretToVariableNameInValue([value], nameStr);
        final valueStr = _emitExpr(value);
        return "$nameStr := $valueStr";
      default:
        throw "Invalid variable mutation $nodes";
    }
  }

  String _emitCondStatement(List<Node> nodes) {
    bool throwIfAnyIfBranchIsNotCorrect(List<Node> branches) {
      for (var branch in branches) {
        switch (branch) {
          case ListNode(contents: [Node(), ...]):
            null;
          default:
            throw "Invalid branch: $branch in cond statement: $nodes";
        }
      }
      return true;
    }

    String emitIfBranches(List<Node> ifBranches) {
      return ifBranches.indexed
          .map((elem) {
            final ListNode(contents: [Node predicate, ...bodyNodes]) =
                elem.$2 as ListNode;
            final keywordStr = elem.$1 > 0 ? "else if" : "if";
            final predicateStr = _emitExpr(predicate);
            final bodyStr = _emitStatementBody(bodyNodes as List<Node>);
            return "$keywordStr ($predicateStr) {\n$bodyStr\n}";
          })
          .join("\n");
    }

    switch (nodes) {
      // (cond (... ...) (... ...) (else ...)) (cond with an else branch)
      case [
            ...final ifBranches,
            ListNode(
              contents: [SymbolNode(contents: "else"), ...final elseBodyNodes],
            ),
          ]
          when throwIfAnyIfBranchIsNotCorrect(ifBranches):
        final ifBranchesStr = emitIfBranches(ifBranches);
        final elseBodyStr = _emitStatementBody(elseBodyNodes as List<Node>);
        return "$ifBranchesStr\nelse {\n$elseBodyStr\n}";
      // (cond (... ...) (... ...)) (cond with only if branches and no else branch)
      case [...final ifBranches]
          when throwIfAnyIfBranchIsNotCorrect(ifBranches):
        final ifBranchesStr = emitIfBranches(ifBranches);
        return ifBranchesStr;
      default:
        throw "Invalid cond statement: $nodes";
    }
  }

  String _emitNegation(List<Node> nodes) {
    switch (nodes) {
      // (not (+ 5 3))
      case [
            ListNode(
              contents: [
                SymbolNode(contents: final op as String),
                ...final restNodes as List<Node>,
              ],
            ),
          ]
          when _operators.containsKey(op):
        final exprStr = _emitOperator(op, restNodes);
        return "!($exprStr)";
      // (not false)
      case [final Node negationNode]:
        final exprStr = _emitExpr(negationNode);
        return "!$exprStr";
      default:
        throw "Invalid negation: $nodes";
    }
  }

  String _emitDoBlock(List<Node> nodes) {
    final blockStr = _emitStatementBody(nodes);
    return blockStr;
  }

  String _emitOperator(String op, List<Node> nodes) {
    String emitArg(Node arg) {
      // (+ 5 3)
      switch (arg) {
        case ListNode(
              contents: [
                SymbolNode(contents: final op as String),
                ...final restNodes as List<Node>,
              ],
            )
            when _operators.containsKey(op):
          final exprStr = _emitOperator(op, restNodes);
          return "($exprStr)";
        // false
        case final Node argNode:
          final exprStr = _emitExpr(argNode);
          return exprStr;
      }
    }

    switch (nodes) {
      case [...final args] when args.length >= 2:
        final opStr = _operators[op]!;
        final argsStr = args.map(emitArg).join(" $opStr ");
        return argsStr;
      default:
        throw "Invalid use of operator $nodes";
    }
  }

  String _emitFunctionCall(SymbolNode name, List<Node> nodes) {
    final nameStr = _emitSymbol(name);
    final argsStr = _emitCommaSeparatedItems(nodes);
    return "$nameStr($argsStr)";
  }

  /*============================ Atomic emissions ============================*/

  String _emitNumber(NumberNode number) {
    return (number.contents as num).toString();
  }

  String _emitSymbol(SymbolNode symbol, {bool private = false}) {
    final validIndentifier = escapeInvalidChars(
      snakeCasify(symbol.contents as String),
    );
    return private ? "_$validIndentifier" : validIndentifier;
  }

  String _emitString(StringNode string) {
    return "\"${string.contents}\"";
  }

  String _emitValueList(ValueListNode valueList) {
    final exprs = _emitCommaSeparatedItems(valueList.contents);
    return "[$exprs]";
  }

  String _emitList(ListNode list) {
    final exprs = list.contents as List<Node>;
    final idNode = exprs[0] as SymbolNode;

    switch (exprs) {
      case [SymbolNode(contents: final id as String), ...final nodes]:
        return switch (id) {
          // (mod IO ...)
          "defmod" => _emitModuleDefinition(nodes),
          // (use IO)
          "use" => _emitModuleImport(nodes),
          // (def[p] puts [str] ...)
          _ when id.startsWith("def") && id.length >= 3 && id.length <= 4 =>
            _emitFunctionDefinition(
              nodes,
              private: id.contains("p"),
              recursive: false,
            ),
          // (defrec[p] puts [str] ...)
          _ when id.startsWith("defrec") && id.length >= 6 && id.length <= 7 =>
            _emitFunctionDefinition(
              nodes,
              private: id.contains("p"),
              recursive: true,
            ),
          // (fn [str] ...)
          "fn" => _emitFunctionDefinition(nodes, lambda: true),
          // (let[mp] x 5)
          _ when id.startsWith("let") && id.length >= 3 && id.length <= 5 =>
            _emitVariable(
              nodes,
              mutable: id.contains("m"),
              private: id.contains("p"),
            ),
          // (set y 8)
          "set" => _emitVariableMutation(nodes),
          // (cond ((x > 5) ... ...) ((true) ... ...) (else ...))
          "cond" => _emitCondStatement(nodes),
          // (not false)
          "not" => _emitNegation(nodes),
          // (do ... ... ...) (groups multiple exprs together for use in (if ... ...) etc.)
          "do" => _emitDoBlock(nodes),
          // (+ x y)
          _ when _operators.containsKey(id) => _emitOperator(id, nodes),
          // anything else (not a keyword)
          _ => _emitFunctionCall(idNode, nodes),
        };
      default:
        throw "Invalid list in _emitList: $list";
    }
  }

  String _emitExpr(Node expr) {
    return switch (expr) {
      NumberNode() => _emitNumber(expr),
      SymbolNode() => _emitSymbol(expr),
      StringNode() => _emitString(expr),
      ValueListNode() => _emitValueList(expr),
      ListNode() => _emitList(expr),
    };
  }

  Future<String> emit(Node ast, {Node? stdlibAst}) async {
    final astWithStdlib =
        stdlibAst != null
            ? (stdlibAst.contents as List<Node>) + (ast.contents as List<Node>)
            : ast.contents as List<Node>;

    final expandedAst = MacroExpander().expand(astWithStdlib);

    final ir = _emitStatementBody(expandedAst);
    print("NONFORMATTED IR: $ir");
    final maybeFormattedIr = await _tryFormatIr(ir);

    return maybeFormattedIr;
  }
}

File: src/Compiler/Emitter.hs
module Compiler.Emitter (emit) where

import Compiler.MacroExpander (expand)
import Compiler.Mod (Node (ListNode, NumberNode, StringNode, SymbolNode, ValueListNode), isListNode, isSymbolNode)
import Data.List (intercalate)
import Data.Maybe (fromJust, isJust)
import Utils (escapeInvalidChars, snakeCasify)

operators :: [(String, String)]
operators =
  [ -- Math
    ("+", "+"),
    ("-", "-"),
    ("*", "*"),
    ("/", "/"),
    -- Equality
    ("<", "<"),
    ("<=", "<="),
    (">", ">"),
    (">=", ">="),
    ("=", "=="),
    ("not=", "!="),
    -- Logical
    ("and", "&&"),
    ("or", "||")
  ]

specialFormsNotTerminatedWithSemicolons :: [String]
specialFormsNotTerminatedWithSemicolons = ["do"]

------------------------- Helper emission functions --------------------------

emitCommaSeparatedItems :: [Node] -> String
emitCommaSeparatedItems list = intercalate ", " $ map emitExpr list

emitStatementBody :: [Node] -> String
emitStatementBody bodyNodes =
  let emitInner node =
        let exprStr = emitExpr node
            identifier = case node of
              ListNode (SymbolNode str : _) -> str
              ValueListNode (SymbolNode str : _) -> str
              _ -> ""
            shouldTerminateWithSemicolon = identifier `notElem` specialFormsNotTerminatedWithSemicolons
         in exprStr ++ (if shouldTerminateWithSemicolon then ";" else "")
   in intercalate "\n" $ map emitInner bodyNodes

------------------ Special form (keyword/syntax) emissions -------------------

emitModuleDefinition :: [Node] -> String
emitModuleDefinition (SymbolNode name : bodyNodes) =
  let nameStr = emitSymbol $ SymbolNode name
      bodyStr = emitStatementBody bodyNodes
   in "module " ++ nameStr ++ " = {\n" ++ bodyStr ++ "\n}"
emitModuleDefinition nodes = error $ "Invalid module definition: " ++ show nodes

emitModuleImport :: [Node] -> String
-- (use List ...) regular module import (variadic)
emitModuleImport names
  | all isSymbolNode names =
      let aux name =
            let nameStr = emitSymbol name
             in "open " ++ nameStr
          importsStr = intercalate "\n" $ map aux names
       in importsStr
-- (use List (sort (cons 0 [1 2 3]))) shorthand for importing a module into one statement
emitModuleImport [SymbolNode name, bodyNode] =
  let nameStr = emitSymbol $ SymbolNode name
      bodyStr = emitExpr bodyNode
   in nameStr ++ ".(" ++ bodyStr ++ ")"
emitModuleImport nodes = error $ "Invalid module import: " ++ show nodes

emitFunctionDefinition :: [Node] -> Bool -> Bool -> String
emitFunctionDefinition nodes recursive lambda =
  let aux name args bodyNodes =
        let nameStr = if null name then "" else emitSymbol $ SymbolNode name
            argsStr = if null args then "" else emitCommaSeparatedItems args
            bodyStr = emitStatementBody bodyNodes
            maybeRecursiveStr = if not lambda && recursive then " rec " else " "
            keywordStr = if not lambda then "let" ++ maybeRecursiveStr ++ nameStr ++ " = " else ""
         in keywordStr ++ "(" ++ argsStr ++ ") => {\n" ++ bodyStr ++ "\n}"
   in case nodes of
        -- Regular function
        SymbolNode name : ValueListNode args : bodyNodes | all isSymbolNode args -> aux name args bodyNodes
        -- Regular function with no argument list shorthand
        SymbolNode name : bodyNodes -> aux name [] bodyNodes
        -- Anonymous/lambda function
        ValueListNode args : bodyNodes | lambda && all isSymbolNode args -> aux "" args bodyNodes
        -- Anonymous/lambda function with argument list shorthand
        bodyNodes | lambda -> aux "" [] bodyNodes
        _ -> error $ "Invalid function declaration: " ++ show nodes

emitLetBinding :: [Node] -> String
emitLetBinding [name, value]
  | not $ isListNode name =
      let nameStr = emitExpr name
          valueStr = emitExpr value
       in "let " ++ nameStr ++ " = " ++ valueStr
emitLetBinding nodes =
  error $ "Invalid let binding: " ++ show nodes

emitCondStatement :: [Node] -> String
emitCondStatement nodes =
  let aux (ListNode (predicate : bodyNodes)) isFirst =
        let keywordStr = if isFirst then "if" else "else if"
            predicateStr = emitExpr predicate
            bodyStr = emitStatementBody bodyNodes
         in keywordStr ++ " (" ++ predicateStr ++ ") {\n" ++ bodyStr ++ "\n}"
      aux branch _ = error $ "Invalid branch in cond statment: " ++ show branch
      isFirstBranch _ [] = error "Invalid branches in isFirstBranch: []"
      isFirstBranch branch [firstBranch] = branch == firstBranch
      isFirstBranch branch (firstBranch : _) = branch == firstBranch
   in case reverse nodes of
        -- (cond (... ...) ... (else ...)) with else branch
        ListNode (SymbolNode "else" : elseBodyNodes) : reversedIfBranches ->
          let ifBranches = reverse reversedIfBranches
              ifBranchesStr = intercalate "\n" $ map (\el -> aux el (isFirstBranch el ifBranches)) ifBranches
              elseBodyStr = emitStatementBody elseBodyNodes
           in ifBranchesStr ++ "\nelse {\n" ++ elseBodyStr ++ "\n}"
        -- (cond (... ...) ...) only if branches no else branch
        reversedIfBranches
          | not $ null reversedIfBranches ->
              let ifBranches = reverse reversedIfBranches
                  ifBranchesStr = intercalate "\n" $ map (\el -> aux el (isFirstBranch el ifBranches)) ifBranches
               in ifBranchesStr
        _ -> error $ "Invalid cond statement: " ++ show nodes

emitNegation :: [Node] -> String
emitNegation [ListNode (SymbolNode operator : restNodes)]
  | isJust $ lookup operator operators =
      let exprStr = emitOperator operator restNodes
       in "!(" ++ exprStr ++ ")"
emitNegation [negationNode] =
  let exprStr = emitExpr negationNode
   in "!" ++ exprStr
emitNegation nodes = error $ "Invalid negation: " ++ show nodes

emitDoBlock :: [Node] -> String
emitDoBlock nodes =
  let bodyStr = emitStatementBody nodes
   in bodyStr

emitOperator :: String -> [Node] -> String
emitOperator operator args
  | isJust (lookup operator operators) && length args >= 2 =
      let aux (ListNode (SymbolNode operator' : restNodes))
            | isJust $ lookup operator' operators =
                let exprStr = emitOperator operator' restNodes
                 in "(" ++ exprStr ++ ")"
          aux argNode =
            let exprStr = emitExpr argNode
             in exprStr
          operatorStr = fromJust $ lookup operator operators
          argsStr = intercalate (" " ++ operatorStr ++ " ") $ map aux args
       in argsStr
emitOperator _ nodes = error $ "Invalid use of operator: " ++ show nodes

emitFunctionCall :: String -> [Node] -> String
emitFunctionCall name nodes =
  let nameStr = emitSymbol $ SymbolNode name
      argsStr = emitCommaSeparatedItems nodes
   in nameStr ++ "(" ++ argsStr ++ ")"

------------------------------ Atomic emissions ------------------------------

emitNumber :: Node -> String
emitNumber (NumberNode num) = show num
emitNumber node = error $ "Invalid number: " ++ show node

emitSymbol :: Node -> String
emitSymbol (SymbolNode sym) = escapeInvalidChars $ snakeCasify sym
emitSymbol node = error $ "Invalid symbol: " ++ show node

emitString :: Node -> String
emitString (StringNode str) = "\"" ++ str ++ "\""
emitString node = error $ "Invalid string: " ++ show node

emitValueList :: Node -> String
emitValueList (ValueListNode nodes) =
  let exprs = emitCommaSeparatedItems nodes
   in "[" ++ exprs ++ "]"
emitValueList node = error $ "Invalid value list: " ++ show node

emitList :: Node -> String
-- (defmod IO ...)
emitList (ListNode (SymbolNode "defmod" : nodes)) = emitModuleDefinition nodes
-- (use IO)
emitList (ListNode (SymbolNode "use" : nodes)) = emitModuleImport nodes
-- (def puts [str] ...)
emitList (ListNode (SymbolNode "def" : nodes)) = emitFunctionDefinition nodes False False
-- (defrec puts [str] ...)
emitList (ListNode (SymbolNode "defrec" : nodes)) = emitFunctionDefinition nodes True False
-- (fn [str] ...)
emitList (ListNode (SymbolNode "fn" : nodes)) = emitFunctionDefinition nodes False True
-- (let x 5)
emitList (ListNode (SymbolNode "let" : nodes)) = emitLetBinding nodes
-- (cond (true ...) (false ...) ... (else ...))
emitList (ListNode (SymbolNode "cond" : nodes)) = emitCondStatement nodes
-- (not true)
emitList (ListNode (SymbolNode "not" : nodes)) = emitNegation nodes
-- (do ...)
emitList (ListNode (SymbolNode "do" : nodes)) = emitDoBlock nodes
-- (+ 5 3)
emitList (ListNode (SymbolNode identifier : nodes)) | isJust $ lookup identifier operators = emitOperator identifier nodes
-- everything else
emitList (ListNode (SymbolNode identifier : nodes)) = emitFunctionCall identifier nodes
emitList exprs = error $ "Invalid list in emitList: " ++ show exprs

emitExpr :: Node -> String
emitExpr (NumberNode num) = emitNumber $ NumberNode num
emitExpr (SymbolNode sym) = emitSymbol $ SymbolNode sym
emitExpr (StringNode str) = emitString $ StringNode str
emitExpr (ValueListNode nodes) = emitValueList $ ValueListNode nodes
emitExpr (ListNode nodes) = emitList $ ListNode nodes

emit :: Node -> String
emit (ListNode astNodes) =
  let expandedAst = expand astNodes
   in emitStatementBody expandedAst
emit _ = error "AST has to be wrapped with a ListNode"